rfcs/testing.md

```diff
@@ -0,0 +1,75 @@
+# RFC: Testing standards
+
+---
+
+_@andrefcdias_
+
+## Summary
+
+The purpose of this RFC is to standardize our testing process and clarify the purpose of each of the testing layers we have. It is supposed to define what is the intended result from tests and why they are required.
+
+## Background
+
+This RFC came from the discussion of unit testing styles in components such as Flex and Text and the debate of having a valuable test vs a "correct" one.
+
+## Problem statement
+
+At the moment, everyone has their own testing standards and there is no common approach to problems like components that only apply styles and need to be tested.
+There is no majority decision on the solution for this problem, being split between Visual Regression tests (VR tests) and Unit Tests (UT).
+
+By defining a common approach, we can make the testing process less ambiguous, as well as provide a guide for the community to follow.
+
+## Detailed Design or Proposal
+
+To start, this is the proposal for defining the goals of each testing layer.
+
+- **Conformance**
+
+  Ensuring the consistency of our components. They test that props like `as`, `ref`, `className` are supported and that the component is rendered.
+
+  These can be achieved simply through the `isConformant` function generated when creating a converged component.
+
+- **Snapshot**
+
+  Stricly DOM structure consistency enforcing. This will guarantee we do not introduce breaking changes for users who expect a certain structure to be rendered by our components.
+
+  They are achieved with Jest's `expect(component).toMatchSnapshot()`.
+
+- **Unit**
+
+  Focus on testing the behaviour or functionality. We should use them to guarantee the outcome, wheter this is a styling, behaviour or functionality change.
+  We should try to focus on behaviour, when possible, but mainly on the expected results for the code.
+
+  These are all the additional unit tests defined in your `*.test.ts[x]` file.
```

Comment:
from experience unit test should be avoided as much as possible with few exceptions:

- non UI related complex helpers that are used in other places

with that said, ideal scenario (for a design system/component/UI) is to author integration like test (enforced by `testing-library` toolkit)
- this creates less brittle tests that gives us good enough guarantees that the functionality of tested UI logic is correct
- usually ends up with less testing changes needed when one adds/removes/fixes features

Where this should happen is up to us to determine
- on artificial environment  / jest + jsdom
- real browser  / cypress + storybook

Reply:
I'm all for these changes but then we need to fit in the style enforcing somewhere else. Should we go forward with only testing styles with VR tests then?

Reply:
> should we go forward with only testing styles with VR tests then?

probably yes for most use cases ?  again if there is some non trivial logic (a mixin for example) for styles that should be tested on "unit/integration" level

Reply:
> from experience unit test should be avoided as much as possible

We might need to clarify some definitions here--of our existing component tests in a `ComponentName.test.tsx` file, what would you consider to be unit tests vs. integration tests or some other category? Maybe this is imprecise but I'm used to thinking of most of those tests as unit tests, and I'm guessing that's not what you mean.

Reply:
> We might need to clarify some definitions here
true that @ecraig12345 !

Here as quick example (for component space context - UI ):

**Unit test:**
- back in the day -> using enzyme -> testing internal stuff like state etc

**Integration test:**
- testing against "real" DOM -> testing-library basically enforces this kind of approach
- testing ideally not "every" component - but only integrated scenarios that make sense

as a recap this article is nice overview of approaches https://kentcdodds.com/blog/write-tests